File: backend/app/services/output_persistence_service.py
import os
import re
from pathlib import Path

_PROJECT_ROOT = Path(__file__).resolve().parents[3]
_DEFAULT_OUTPUT_DIR = _PROJECT_ROOT / "tailored-resumes"
OUTPUT_DIR = Path(os.getenv("TAILORED_RESUMES_DIR", _DEFAULT_OUTPUT_DIR)).resolve()

_INVALID_FILENAME_CHARS = re.compile(r"[^A-Za-z0-9.\-_ ()]+")


def _sanitize_filename(filename: str) -> str:
    cleaned = filename.replace("/", "-").replace("\\", "-").strip()
    cleaned = _INVALID_FILENAME_CHARS.sub("", cleaned)
    if not cleaned:
        cleaned = "Resume.docx"
    if not cleaned.lower().endswith(".docx"):
        cleaned = f"{cleaned}.docx"
    return cleaned


def _next_available_path(base_dir: Path, filename: str) -> Path:
    candidate = base_dir / filename
    if not candidate.exists():
        return candidate

    stem = candidate.stem
    suffix = candidate.suffix
    idx = 1
    while True:
        retry_candidate = base_dir / f"{stem}-{idx}{suffix}"
        if not retry_candidate.exists():
            return retry_candidate
        idx += 1


def persist_resume_output(doc_bytes: bytes, suggested_filename: str) -> tuple[Path, str]:
    OUTPUT_DIR.mkdir(parents=True, exist_ok=True)
    safe_filename = _sanitize_filename(suggested_filename)
    target_path = _next_available_path(OUTPUT_DIR, safe_filename)
    target_path.write_bytes(doc_bytes)

    try:
        relative_output_path = str(target_path.relative_to(_PROJECT_ROOT))
    except ValueError:
        relative_output_path = str(target_path)

    return target_path, relative_output_path

File: server/src/services/outputPersistenceService.ts
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const PROJECT_ROOT = path.join(__dirname, '../../../../');
const DEFAULT_OUTPUT_DIR = path.join(PROJECT_ROOT, 'tailored-resumes');
const OUTPUT_DIR = process.env.TAILORED_RESUMES_DIR || DEFAULT_OUTPUT_DIR;

const INVALID_CHARS = /[^A-Za-z0-9.\-_ ()]+/g;

function sanitizeFilename(filename: string): string {
  let cleaned = filename.replace(/\//g, '-').replace(/\\/g, '-').trim();
  cleaned = cleaned.replace(INVALID_CHARS, '');
  if (!cleaned) cleaned = 'Resume.docx';
  if (!cleaned.toLowerCase().endsWith('.docx')) cleaned = `${cleaned}.docx`;
  return cleaned;
}

function nextAvailablePath(dir: string, filename: string): string {
  let candidate = path.join(dir, filename);
  if (!fs.existsSync(candidate)) return candidate;

  const ext = path.extname(filename);
  const stem = path.basename(filename, ext);
  let idx = 1;
  while (true) {
    candidate = path.join(dir, `${stem}-${idx}${ext}`);
    if (!fs.existsSync(candidate)) return candidate;
    idx++;
  }
}

export function persistResumeOutput(docBytes: Buffer, suggestedFilename: string): [string, string] {
  fs.mkdirSync(OUTPUT_DIR, { recursive: true });
  const safeFilename = sanitizeFilename(suggestedFilename);
  const targetPath = nextAvailablePath(OUTPUT_DIR, safeFilename);
  fs.writeFileSync(targetPath, docBytes);

  let relativePath: string;
  try {
    relativePath = path.relative(PROJECT_ROOT, targetPath);
  } catch {
    relativePath = targetPath;
  }

  return [targetPath, relativePath];
}
